doc/sphinx/language/extensions/compil-steps.rst

```diff
@@ -0,0 +1,169 @@
+
+* During :gdef:`internalization` a number of things are resolved. This
+  includes :ref:`name resolution <qualified-names>`, :term:`notation
+  interpretation` and introduction of :term:`holes <hole>` for :ref:`implicit
+  arguments <ImplicitArguments>`.
+  :gdef:`Notation interpretation <notation interpretation>`
+  translates each syntactic element to a term,
+  for instance :n:`1` can be interpreted as the
+  natural number :n:`S O` then :n:`2` is interpreted as :n:`S (S O)`,
+  then :n:`2 * 2` as :n:`Nat.mul (S (S O)) (S (S O))` and finally our
+  whole term as :n:`Nat.add (S O) (Nat.mul (S (S O)) (S (S O)))`. The
+  same expression can be given multiple interpretations in various
+  contexts thanks to :ref:`Scopes`.
+
+* Finally, :gdef:`type inference`, can use the various mechanisms described in
+  this section to fill gaps (for instance with :ref:`canonical structures
+  <canonicalstructures>` or :ref:`typeclasses`) or fix terms (for
+  instance with :ref:`coercions <coercions>`) to obtain fully detailed terms in
+  the :ref:`Core Language <core-language>`.
+
+For each term, Rocq performs these steps *successively* and
+*independently*.
+Then, the result goes through the type checking phases discussed in
+:ref:`previous chapter <core-language>`.
+None of the steps has any impact on the previous ones.
+In particular, no typecheking is involved during parsing or
+internalization. Also note that none of the features resolved during
+these phases, like unqualified names, implicit arguments or notations,
+remains during the later type inference and type checking phases.
+
+.. note::
+
+   The :term:`type inference` phase together with, all or part of, the
+   previous steps is sometimes called elaboration in the literature.
+
+.. example:: Simple interleaving of intern and type inference phases
+
+   The command :g:`Definition foo : T := body.` has a trivial
+   :term:`synterp` phase. Indeed, it doesn't influence any further
+   parsing. Its :term:`interp` phase will internalize :g:`T` and infer
+   types in it, then it will internalize :g:`body` and infer types in
+   it, using :g:`T` as expected type. Note that the result of type
+   inference in :g:`T` matters in internalization of :g:`body`, for
+   instance for selecting notation scopes. Finally, the resulting
+   :g:`foo : T := body` is sent to the kernel.
+
+.. example:: Delayed steps
+
+   :g:`Ltac foo := exact term.` will internalize :g:`term` and save
+   the result. Only when the :g:`foo` tactic will be called, will the
+   type inference on the resulting :n:`term` be run, with the type of
+   the current goal as expected type. If this succeeds, the proof
+   state will be updated to fill the goal, and the kernel will see the
+   result at :cmd:`Qed`.
+
+.. example:: Reserved notation
+
+   The command :g:`Reserved Notation "x + y" (at level 50, left
+   associativity).` has a non trivial synterp phase, as it extends the
+   parser so that :n:`_ + _` can later be parsed.
```

Comment:
Is it worth mentioning that it has a trivial `interp` phase?

Reply:
Why not, done